### PR TITLE
Focus the currently focused node when it mounts

### LINKED ISSUE #591: Inserting a root node does not update focus correctly
**Describe the bug**
Inserting `a b c` as root nodes - anywhere in the document - should put focus on `c` when the editor is re-rendered. Instead, focus is lost.

**To Reproduce**

1. Load the default editor, and switch to block mode
2. Place the cursor on the line beneath `quuz`, and type `a b c`.
3. Hit enter

**Expected behavior**
`a`, `b`, and `c` appear as root nodes, with `c` activated

**Desktop (please complete the following information):**

- OS: OS X Catalina
- Browser: Brave
- Version 1.31 (chromium 95)

**Additional context**
This would make a good test case for #468

### PR DESCRIPTION
**Pull request by pcardune**
Fixes #591

This ended up being surprisingly complicated. The solution to the actual problem was simple: add a useEffect() that focuses the currently focused node when it mounts. Doing this worked fine and all the tests passed, except there was a cryptic react warning message that showed up when the tests ran. The warning only showed up in tests and not when doing exactly the same thing in the browser.

The issue has to do with the timing of calls to `focus(),` though not the call to `focus()` that I added, but rather one that's been in there already for a while, and which is deeply buried in codemirror in the call to `setBookmark()`. I don't know why the warnings didn't show up until I made this change. In any case, you are theoretically not supposed to monkey with the dom (e.g. by calling `focus()`) while react is rendering, and instead use `useEffect()` to wait until rendering is finished, but we were calling `setBookmark()` while rendering `ToplevelBlockEditable`.

Anyway, to fix the issue, I moved the call to `setBookmark()` to happen inside a `useEffect()`, which resulted in a different problem. The order in which `useEffect()` callbacks are called is depth first. So the `useEffect()` inside `NodeEditable` which selects the text as soon as the NodeEditable component mounts was being called _before_ it's parent component's `useEffect()` had called `setBookmark()`, so text could node be selected because it wasn't in the document yet.

To avoid the troubles with using `setAfterDOMUpdate` or similar to "wait until react is done", I had to move responsibility for calling selectElement() up the component hierarchy to `ToplevelBlockEditable`. So now we only select the element text _after_ it's container element has been inserted into the document via `setBookmark()`. It's the first time I've had the chance to use [`useImperativeHandle()`](https://reactjs.org/docs/hooks-reference.html#useimperativehandle).

codemirror really makes things 10x more complicated than they would otherwise be, but I hope the comments I've left behind are explanatory enough.

### CODE REVIEW

**Comment by pcardune** [comment]
This also fixes #592

**Comment by schanzer** [comment]
@pcardune "add a useEffect() that focuses the currently focused node when it mounts"

This is not the desired behavior. If a blind user has focus on the "switch to blocks mode" button, they shouldn't have their focus jump to the first block of the editor. The only time focus should jump like that is when they're _already in the block editor_, and they make an edit that switches focus.

File: packages/codemirror-blocks/src/components/Node.tsx [hunk 1]
```diff
@@ -240,6 +240,15 @@ const Node = ({ expandable = true, ...props }: Props) => {
...
 
+  // If this is the currently focused node, then focus the node element
+  // as soon as we finish rendering.
+  const focusedNode = useSelector(selectors.getFocusedNode);
+  useEffect(() => {
+    if (nodeElementRef.current && focusedNode?.id === props.node.id) {
+      nodeElementRef.current?.focus();
+    }
+  }, [focusedNode?.id, props.node.id]);
+
```
Comment (schanzer) [inline packages/codemirror-blocks/src/components/Node.tsx:251]: I feel like this might be the only thing that needs to change?

**Comment by pcardune** [comment]
How about this?

File: packages/codemirror-blocks/src/components/Node.tsx (hunk 1 above)
Comment (schanzer) [inline packages/codemirror-blocks/src/components/Node.tsx:261]: My understanding of line 250 is that this hook "subscribes" to the `focusedNode` selector and the node `id`. Does this mean that any time the `focusedNode` changes, _every_ node asks if the focused node's id is the same as the node's id? If not, what does it mean?

**Comment by schanzer** [comment]
Perfect! Merging now. Thanks for adding the tests!

File: packages/codemirror-blocks/src/components/Node.tsx (hunk 1 above)
Comment (pcardune) [reply packages/codemirror-blocks/src/components/Node.tsx:261]: yep, that's what this means.

**Pull request merged**